sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/RetryUtil.java

```diff
@@ -51,51 +56,54 @@ public static AmqpRetryPolicy getRetryPolicy(AmqpRetryOptions options) {
-     *         Otherwise, propagates a {@link TimeoutException}.
+     *     Otherwise, propagates a {@link TimeoutException}.
      */
-    public static <T> Flux<T> withRetry(Flux<T> source, Duration operationTimeout, AmqpRetryPolicy retryPolicy) {
-        return Flux.defer(() -> source.timeout(operationTimeout))
-            .retryWhen(Retry.withThrowable(errors -> retry(errors, retryPolicy)));
+    public static <T> Flux<T> withRetry(Flux<T> source, AmqpRetryOptions retryOptions, String timeoutMessage) {
+        return source.timeout(retryOptions.getTryTimeout())
+            .retryWhen(createRetry(retryOptions))
+            .doOnError(error -> LOGGER.error(timeoutMessage, error));
     }
 
     /**
      * Given a {@link Mono} will apply the retry policy to it when the operation times out.
      *
      * @param source The publisher to apply the retry policy to.
+     *
      * @return A publisher that returns the results of the {@link Flux} if any of the retry attempts are successful.
-     *         Otherwise, propagates a {@link TimeoutException}.
+     *     Otherwise, propagates a {@link TimeoutException}.
      */
-    public static <T> Mono<T> withRetry(Mono<T> source, Duration operationTimeout, AmqpRetryPolicy retryPolicy) {
-        return Mono.defer(() -> source.timeout(operationTimeout))
-            .retryWhen(Retry.withThrowable(errors -> retry(errors, retryPolicy)));
+    public static <T> Mono<T> withRetry(Mono<T> source, AmqpRetryOptions retryOptions, String timeoutMessage) {
+        return source.timeout(retryOptions.getTryTimeout())
+            .retryWhen(createRetry(retryOptions))
+            .doOnError(error -> LOGGER.error(timeoutMessage, error));
     }
 
-    private static Flux<Long> retry(Flux<Throwable> source, AmqpRetryPolicy retryPolicy) {
-        return source.zipWith(Flux.range(1, retryPolicy.getMaxRetries() + 1),
-            (error, attempt) -> {
-                if (attempt > retryPolicy.getMaxRetries()) {
-                    LOGGER.warning("Retry attempts are exhausted. Current: {}. Max: {}.", attempt,
-                        retryPolicy.getMaxRetries());
-                    throw Exceptions.propagate(error);
-                }
-
-                if (error instanceof TimeoutException) {
-                    LOGGER.info("TimeoutException error occurred. Retrying operation. Attempt: {}.",
-                        attempt, error);
-
-                    return retryPolicy.calculateRetryDelay(error, attempt);
-                } else if (error instanceof AmqpException && (((AmqpException) error).isTransient())) {
-                    LOGGER.info("Retryable error occurred. Retrying operation. Attempt: {}. Error condition: {}",
-                        attempt, ((AmqpException) error).getErrorCondition(), error);
+    static Retry createRetry(AmqpRetryOptions options) {
+        if (options.getDelay() == null || options.getDelay().equals(Duration.ZERO)) {
```

Comment:
Based on the previous topic of no delay or zero delay resulting in no retrying, should we also check if the delay is negative?